File: c/ofi_core.c
#include "ofi_core.h"

#include <limits.h>

int32_t ofi_sat32(int64_t x) {
    if (x > INT32_MAX) return INT32_MAX;
    if (x < INT32_MIN) return INT32_MIN;
    return (int32_t)x;
}

void ofi_core_init(ofi_core_t *core, unsigned k_decay, uint32_t theta_q16) {
    core->k = k_decay;
    core->theta_q16 = theta_q16;
    core->has_prev = false;
    core->acc = 0;
    core->depth_acc = 0;
}

/* Spec section 4.2 rule 3: shifts are arithmetic and round toward negative
 * infinity. On every mainstream compiler (gcc, clang, MSVC) `>>` on a
 * signed integer type is already arithmetic, matching Python's `>>` on
 * negative ints -- this is documented per spec's requirement rather than
 * worked around, since no target compiler for this project behaves
 * otherwise. */
static int64_t arith_shr(int64_t x, unsigned k) {
    return x >> k;
}

ofi_result_t ofi_core_event(ofi_core_t *core, ofi_top_t t) {
    ofi_result_t r = {0, 0, 0};

    if (!core->has_prev) {
        core->prev = t;
        core->has_prev = true;
        core->depth_acc = ((int64_t)t.bid_qty + (int64_t)t.ask_qty) << core->k;
        return r; /* (0, 0, 0): seed only, not a signal */
    }

    ofi_top_t p = core->prev;

    /* --- OFI event term: 4 compares, 4 conditional adds --- */
    int64_t e = 0;
    if (t.bid_px >= p.bid_px) e += t.bid_qty;
    if (t.bid_px <= p.bid_px) e -= p.bid_qty;
    if (t.ask_px <= p.ask_px) e -= t.ask_qty;
    if (t.ask_px >= p.ask_px) e += p.ask_qty;
    int32_t e_sat = ofi_sat32(e);

    /* --- EWMA accumulate: one shift, two adds --- */
    core->acc = core->acc - arith_shr(core->acc, core->k) + e_sat;

    /* --- depth EWMA, same structure --- */
    int64_t d = (int64_t)t.bid_qty + (int64_t)t.ask_qty;
    core->depth_acc = core->depth_acc - arith_shr(core->depth_acc, core->k) + d;

    /* --- threshold WITHOUT division (spec section 3.2) --- */
    int64_t lhs = core->acc << 16;
    int64_t rhs = arith_shr((int64_t)core->theta_q16 * core->depth_acc, core->k);

    uint8_t flags = 0;
    if (lhs > rhs) flags |= 0x1u;   /* buy pressure */
    if (lhs < -rhs) flags |= 0x2u;  /* sell pressure */

    core->prev = t;

    r.e_n = e_sat;
    r.acc = core->acc;
    r.flags = flags;
    return r;
}

File: src/ofi_signal/ofi_core.py
"""Bit-exact integer OFI with EWMA accumulation. Ported near-verbatim from
ofi_golden_model_spec.md section 4.3 -- this is the file that must match
the RTL cycle-for-cycle, so do not "clean it up" with anything that isn't
also in the spec (no floats, no implicit division, explicit saturation).
"""

from __future__ import annotations

from dataclasses import dataclass

INT32_MIN, INT32_MAX = -(1 << 31), (1 << 31) - 1


def sat32(x: int) -> int:
    return INT32_MAX if x > INT32_MAX else INT32_MIN if x < INT32_MIN else x


@dataclass
class Top:
    bid_px: int  # ticks
    bid_qty: int  # shares
    ask_px: int
    ask_qty: int


class OFICore:
    """Mirrors the RTL exactly: no floats, explicit saturation,
    arithmetic right shift for decay."""

    def __init__(self, k_decay: int, theta_q16: int):
        self.k = k_decay  # EWMA: acc -= acc >> k
        self.theta = theta_q16  # Q16.16 threshold multiplier
        self.prev: Top | None = None
        self.acc = 0  # int48/int64 in RTL
        self.depth_acc = 0  # running depth EWMA
        self.valid = False  # gated on book validity

    def event(self, t: Top) -> tuple[int, int, int]:
        """Process one book state. Returns (e_n, acc, flags)."""
        p = self.prev
        if p is None:
            self.prev = t
            self.depth_acc = (t.bid_qty + t.ask_qty) << self.k
            return (0, 0, 0)

        # --- OFI event term: 4 compares, 4 conditional adds ---
        e = 0
        if t.bid_px >= p.bid_px:
            e += t.bid_qty
        if t.bid_px <= p.bid_px:
            e -= p.bid_qty
        if t.ask_px <= p.ask_px:
            e -= t.ask_qty
        if t.ask_px >= p.ask_px:
            e += p.ask_qty
        e = sat32(e)

        # --- EWMA accumulate: one shift, two adds ---
        self.acc = self.acc - (self.acc >> self.k) + e

        # --- depth EWMA, same structure ---
        d = t.bid_qty + t.ask_qty
        self.depth_acc = self.depth_acc - (self.depth_acc >> self.k) + d

        # --- threshold WITHOUT division (spec section 3.2) ---
        # want: acc / depth > theta   <=>   acc << 16 > theta * depth
        lhs = self.acc << 16
        rhs = (self.theta * self.depth_acc) >> self.k
        flags = 0
        if lhs > rhs:
            flags |= 0b01  # buy pressure
        if lhs < -rhs:
            flags |= 0b10  # sell pressure

        self.prev = t
        return (e, self.acc, flags)
